File: lib/pane/layout/templates/top_split_bottom.ex
defmodule Pane.Layout.Templates.TopSplitBottom do
  @moduledoc """
  A layout with a large top pane (60% by default) and two equal bottom panes (40% total).

  Pane names:
  - :top
  - :bottomLeft
  - :bottomRight
  """

  @behaviour Pane.Layout.Template

  @impl true
  def pane_names, do: [:top, :bottomLeft, :bottomRight]

  @impl true
  def apply(window_target, cwd, options) do
    # Extract options with defaults
    top_size = Map.get(options, :topSize, 60)
    bottom_size = 100 - top_size

    # Generate commands
    commands = []

    # First pane is created by default with the window, so we just need to reference it
    top_target = "#{window_target}.0"
    
    # Split window for bottom section (top/bottom split)
    # Note: In tmux, -v means vertical split, which divides top/bottom
    bottom_left_cmd = Pane.Tmux.Pane.split(
      direction: :vertical, 
      percent: bottom_size, 
      target: top_target,
      cwd: cwd
    )
    commands = commands ++ [bottom_left_cmd]
    bottom_left_target = "#{window_target}.1"
    
    # Split the bottom pane horizontally (left/right split)
    # Note: In tmux, -h means horizontal split, which divides left/right
    bottom_right_cmd = Pane.Tmux.Pane.split(
      direction: :horizontal, 
      percent: 50, 
      target: bottom_left_target,
      cwd: cwd
    )
    commands = commands ++ [bottom_right_cmd]
    bottom_right_target = "#{window_target}.2"
    
    # Select the top pane again to make it active
    select_top_cmd = Pane.Tmux.Pane.select(top_target)
    commands = commands ++ [select_top_cmd]
    
    # Return commands and a map of pane positions to targets
    {
      commands,
      %{
        top: top_target,
        bottomLeft: bottom_left_target,
        bottomRight: bottom_right_target
      }
    }
  end
end

File: nodejs/src/lib/templates/topsplitbottom.js
/**
 * A layout with a large top pane (60%) and two equal bottom panes (40% total).
 * 
 * Pane names:
 * - top
 * - bottomLeft
 * - bottomRight
 */
const { TmuxPane } = require('../tmux');

/**
 * Returns the list of pane position names defined by this template.
 * 
 * @returns {string[]} Array of pane names
 */
function paneNames() {
  return ['top', 'bottomLeft', 'bottomRight'];
}

/**
 * Creates the pane structure within a window.
 * 
 * @param {string} windowTarget - The target identifier for the window
 * @param {string} cwd - The current working directory 
 * @param {Object} options - Additional options
 * @returns {Object} Commands and pane targets
 */
function apply(windowTarget, cwd, options = {}) {
  // First pane is created by default with the window
  const topTarget = `${windowTarget}.0`;
  
  // Create a bottom pane that's 40% of the window height
  const splitCommand1 = TmuxPane.split({ 
    target: topTarget, 
    vertical: true, 
    percentage: 40,
    cwd
  });
  
  // The new pane becomes pane 1
  const bottomLeftTarget = `${windowTarget}.1`;
  
  // Create a bottom right pane by splitting the bottom left pane horizontally
  const splitCommand2 = TmuxPane.split({ 
    target: bottomLeftTarget, 
    horizontal: true, 
    percentage: 50,
    cwd
  });
  
  // The new pane becomes pane 2
  const bottomRightTarget = `${windowTarget}.2`;
  
  // Select the top pane to make it active
  const selectCommand = TmuxPane.select(topTarget);
  
  // Return commands and pane targets
  return {
    commands: [splitCommand1, splitCommand2, selectCommand],
    paneTargets: {
      top: topTarget,
      bottomLeft: bottomLeftTarget,
      bottomRight: bottomRightTarget
    }
  };
}

module.exports = {
  paneNames,
  apply
};